internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,256 @@
+	// Remove old pullers
+	g.pullers.RemoveOld(pl)
+
+	// Add new pullers
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		g.pullers.Add(service.Name, buffer, g.entryCh)
+	}
+}
+
+// serviceStarted is called by the LogManager on the start of a service which
+// logs to this gatherer's target.
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullers.Add(service.Name, buffer, g.entryCh)
+}
+
+// The main control loop for the logGatherer. loop receives logs from the
+// pullers on entryCh, and writes them to the client. It also flushes the
+// client periodically, and exits when the gatherer's tomb is killed.
+func (g *logGatherer) loop() error {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+mainLoop:
+	for {
+		select {
+		case <-g.tomb.Dying():
+			break mainLoop
+
+		case <-ticker.C:
+			// Timeout - flush
+			err := g.client.Flush(g.clientCtx)
+			if err != nil {
+				logger.Noticef("sending logs to target %q: %v", g.targetName, err)
+			}
+
+		case entry := <-g.entryCh:
+			err := g.client.Write(g.clientCtx, entry)
+			if err != nil {
+				logger.Noticef("writing logs to target %q: %v", g.targetName, err)
+			}
+		}
+	}
+
+	// Final flush to send any remaining logs buffered in the client
+	ctx, cancel := context.WithTimeout(context.Background(), g.timeoutFinalFlush)
+	defer cancel()
+	err := g.client.Flush(ctx)
+	if err != nil {
+		logger.Noticef("sending logs to target %q: %v", g.targetName, err)
+	}
+	return nil
+}
+
+// stop tears down the gatherer and associated resources (pullers, client).
+// This method will block until gatherer teardown is complete.
+//
+// The teardown process has several steps:
+//   - If the main loop is in the middle of a flush when we call stop, this
+//     will block the pullers from sending logs to the gatherer. Hence, wait
+//     for the current flush to complete.
+//   - Wait for the pullers to pull the final logs from the iterator.
+//   - Kill the main loop.
+//   - Flush out any final logs buffered in the client.
+func (g *logGatherer) stop() {
+	// Wait up to timeoutCurrentFlush for the current flush to complete (if any)
+	time.AfterFunc(timeoutCurrentFlush, g.clientCancel)
+
+	// Wait up to timeoutPullers for the pullers to pull the final logs from the
+	// iterator and send to the main loop.
+	time.AfterFunc(timeoutPullers, g.pullers.KillAll)
+
+	// Kill the main loop once either:
+	// - all the pullers are done
+	// - timeoutMainLoop has passed
+	select {
+	case <-g.pullers.Done():
+	case <-time.After(timeoutMainLoop):
+	}
+
+	_ = g.tomb.Killf("gatherer stopped")
+	// Wait for final flush in the main loop
+	_ = g.tomb.Wait()
```

Comment:
Why are we ignoring all errors?  Shouldn't this be reporting any errors found so that there's a way out for them?  Among other things, that's nice for example when you want to test that the logic is generally sane, as in tests for example, but it's often also a hint that the logic knows how to die properly. For instance: if all errors are currently ignored, how do we detect and recover from a situation where a network connection dies?  Do we have good recovery for this ATM?

Reply:
This is during teardown, so if something goes wrong, we just have to move on. We can't really block Pebble shutdown because of a bad network connection. It's "best effort" here - try to send off the last logs, but if we can't, we just have to drop them. I think it's fair to log the error to the user though.